### ISSUE TITLE
Add support for initoverlay

### ISSUE THREAD

**Issue by cgwalters** [ostreedev/ostree#3066]
Splitting this from https://github.com/ostreedev/ostree/issues/2867#issuecomment-1744830868
which was inspired by https://github.com/ostreedev/ostree/issues/2867#issuecomment-1602888529

Basically a major flaw with initramfs (whether baked into the kernel binary or separate) is that it's not lazy - the entire thing must be decompressed and parsed before it executes at all.

In this model, we aim to shrink the role of the initramfs to be very small - just enough to mount e.g. `/boot` or `/efi` which contains...a directory tree or an erofs blob say.  The role of the initramfs is to load this image, verify its integrity, and then switch root to it, before finally switching to the real root.

**Comment by cgwalters** [comment]
This would be a new thing that would need to live alongside the kernel data (like initramfs and devicetree).

**Comment by ericcurtin** [comment]
In the UEFI case you end up with tuples of kernel, dtb, initramfs, initoverlayfs etc. in /boot :

/boot/initramfs-0-rescue-c974db5ab26042cbabdaf72bf9433dc7.img: regular file, no read permission
/boot/initoverlayfs-6.4.15-100.fc37.x86_64.img:                EROFS filesystem, compat: SB_CHKSUM MTIME, blocksize=12, exslots=0, uuid=D34116F7-7068-3D4A-9EDC-3FFE23733519
/boot/initramfs-6.4.15-100.fc37.x86_64.img:                     regular file, no read permission

**Comment by cgwalters** [comment]
That's a non-ostree image you're generating though right in that example?  I'd expect to see `/boot/ostree` otherwise.

**Comment by cgwalters** [comment]
On the ostree side I guess I'd vote that we just grab `/usr/lib/ostree-boot/initoverlay.img` to start; we don't care what's in it, we just add it into the kernel-lifecycled data we already have and copy/reflink-if-possible it in to `/boot/ostree`.

How are you thinking signatures on this would work?  I think my strawman proposal here is that we aim to use composefs for this as well.

**Comment by ericcurtin** [comment]
> That's a non-ostree image you're generating though right in that example? I'd expect to see `/boot/ostree` otherwise.

Yes it was, for a start the goals will be to make this work for the following variants:

- ostree UEFI/grub/etc.
- non-ostree UEFI/grub/etc.
- ostree aboot
- non-ostree aboot

It was simpler to do a PoC without ostree to start, you need to run dracut twice for example, etc. and extract one of the initramfs's to an erofs file and this all needs hasn't been coded in ostree yet.

But it does "just work" when I booted it in Fedora, systemd and all the related tools just behave like they would in a normal initramfs except they are running in an initoverlayfs. And that was "fat" Fedora workstation with all the UI stuff, etc.

> On the ostree side I guess I'd vote that we just grab `/usr/lib/ostree-boot/initoverlay.img` to start; we don't care what's in it, we just add it into the kernel-lifecycled data we already have and copy/reflink-if-possible it in to `/boot/ostree`.
> 
> How are you thinking signatures on this would work? I think my strawman proposal here is that we aim to use composefs for this as well.

I did briefly discuss this with @alexlarsson , I was even thinking of changing the name of the project to "initcomposefs" at one point and take advantage of combined engineering efforts in the composefs community. But we came to the conclusion composefs isn't worth it, all you have to do if fs-verity a single initoverlayfs file on boot and we don't care about metadata and such, so simple fs-verity might make more sense (or dm-verity if you want to put the erofs in a partition, but files are nicer to work with).

> BTW one major value of this isn't just boot speed - it means that in the case of e.g. just a kernel security update (and the userspace didn't change) we don't need to ship a whole new initramfs image, which is a really not-small improvement.

Yup there's a few advantages. I have noticed from reading around in the last few days the idea of multiple initramfs's in a single boot isn't a new idea (it's present in some Android implementations to split GKI stuff from other stuff), but this implementation in my biased opinion is superior from a perf and scale perspective.

**Comment by ericcurtin** [comment]
One explicit goal I like to communicate though is we start an init process called pre-init that exec's systemd. The goal is not to replace systemd. systemd, etc. should still do all the heavy lifting including the ostree stuff. It's only role is to do the absolute minimum to switch to initoverlayfs and exec systemd.

**Comment by ericcurtin** [comment]
Related MR https://github.com/coreos/rpm-ostree/pull/4721 

We currently run a script called initoverlayfs-install whenever dracut is called to build an initrd.

**Comment by cgwalters** [comment]
We have threads on this in a few places, but taking this bit here: It seems like a direction initoverlayfs is taking is to effectively become a dracut wrapper.  That approach broadly makes sense to me.

Today, ostree owns updating the kernel+initramfs (+devicetree), and it seems like initoverlayfs here would just become yet another file.  Pretty straightforward.

The thing here is though (related to https://github.com/containers/initoverlayfs/discussions/69 ) is that in the ostree model, the kernel/initramfs are in `/usr/lib/modules/$kver` in the container/commit, but at deployment time are copied into `/boot/ostree`.  (However, if `/boot` is the same filesystem as `/` then this is just a hardlink, so it's free)

However, the ultimate filename includes a sha256 digest (to do deduplication).  This is going to be tricky to deal with...
